### PR TITLE
[WIP] Google Answer block that uses SerpApi to get answers to specific questions.

### PR DESCRIPTION
**Pull request by aickin**
This is a work-in-progress pull request of a new block type to use Google's search page ability to answer specific factual questions. This PR is just on the Rust/CLI side of things; I haven't yet looked at the API or the UI. To use it you add a block to your `.dust` file that looks like:

```
google_answer GOOGLE_ANSWER {
  question: When was ${INPUT.person} born?
}
```

The output of the block is a JSON object of the form:

```json
{
  "question": "When was Anton Chekhov born?"
  "answer": "January 29, 1860"
}
```

The `question` field will give the text of the question that was asked to Google, with all variable substitutions completed. The `answer` field will have the big bold text that Google thinks is the answer. If Google does not have an answer box on the search page for this question, then the `answer` field will be null.

The block requires an environment or Credentials variable called SERP_API_KEY which contains a working API key to SerpApi.

I learned a lot about how Dust works currently and how it's architected by working on this, but I'm certain I've made some silly errors. I'm not as experienced in Rust as in many other languages, and having model code for other blocks to look at was quite helpful.

I'm not filing this PR to necessarily ask for the code to be accepted into main (although I'm happy if you want to do that). I did want to start a discussion, though, about whether this is a reasonable direction for Dust to go. I'm planning to explore adding this to the API and UI next, and I'll update the PR as I work on that. 

Happy to hear any feedback, and congrats on the project so far!

### CODE REVIEW

File: core/src/blocks/google_answer.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+use urlencoding::encode;
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone)]
+pub struct GoogleAnswer {
+    question: String,
+    // hash: String,
+}
+
+impl GoogleAnswer {
```
Comment (spolu) [inline core/src/blocks/google_answer.rs:20]: What's the motivation to go just for the answer? I think we want the full experience no and get the raw SERPAPI reply?

**Review by spolu — COMMENTED** [review]
This looks amazing. I think we really want the full response from SerpApi instead of just the answer so that the block can be as versatile as possible, including search types etc... eventually?

**Comment by spolu** [comment]
Note that @mauriyouth is working on a very similar block here: https://github.com/dust-tt/dust/pull/23

File: core/src/blocks/google_answer.rs (hunk 1 above)
Comment (spolu) [reply core/src/blocks/google_answer.rs:20]: Yes what you propose seems ideal.

Proposal (cc @mauriyouth)
- I take over both your PR to merge them together this afternoon. Your SerpAPI integration, @mauriyouth work to link it to the frontend
- v0 is a raw JSON out of Serp
- From there we iterate to add options to that block to provide more direct use-cases with less code to write?

wdyt?

Comment (mauriyouth) [reply core/src/blocks/google_answer.rs:20]: good for me 👍

**Comment by spolu** [comment]
Also added you both as collaborators on the repo so that you can clone it and PR to it direclty if you wish

File: core/src/blocks/google_answer.rs (hunk 1 above)
Comment (aickin) [reply core/src/blocks/google_answer.rs:20]: I'm already working on a front-end of the Answers version, have it mostly working (still messing with styles, allowing direct enter of SerpApi key on the Dust app page, and a bug that is unearthed in the LLM block UI by adding a non-LLM provider). Do you want me to finish that up in the next couple hours before you take this over? 
image

**Comment by spolu** [comment]
superseded by #25 \o/

**Pull request closed without merging**